### ISSUE TITLE
Non idempotent behavior with subscription

### ISSUE THREAD

**Issue by SaulBerrenson** [nats-io/nats-server#2876]
## Non idempotent behavior with subscription on any client (tested .net, golang, c client)

Hi, when make subscription anybody expect successful or unsuccessful subscription, but if server has slower perfomance than client - you will get non idempotent behavior. When another client after succesful subscription send request with timeout - expecting get or reply or timeout expire, but you will get "no responders available for request" and it's wrong. Because if you subscribe it's meaning operation subscription was completed fully, but not part of.

#### Versions of `nats-server` and affected client libraries used:
2.7.2 and lower

#### OS/Container environment:
Windows x64, Unix x64

#### Steps or code to reproduce the issue:
example code at test for golang (just make 1000 or 10000 iteration with go test -count 10000
We tried at 2 clients (natsc and golang - fail)
 
```
package main

import (
	"github.com/nats-io/nats.go"
	"log"
	"testing"
	"time"
)

func TestSubscription(t *testing.T) {

	nc, err := nats.Connect("localhost:4222")
	if err != nil {
		log.Fatal(err)
	}
	defer nc.Close()

	nsc, err := nats.Connect("localhost:4222")
	if err != nil {
		log.Fatal(err)
	}
	defer nsc.Close()

	subs, _ := nsc.QueueSubscribe("time", "sefsdfgdsgdsgdsf", func(msg *nats.Msg) {
		log.Print(string(msg.Data))
		msg.Respond([]byte("1111"))
	})

	defer subs.Unsubscribe()

	// Send the request
	msg, err := nc.Request("time", nil, time.Second)
	if err != nil {
		log.Fatal(err)
	}

	// Use the response
	log.Printf("Reply: %s", msg.Data)

	// Close the connection
	nc.Close()
	// [end request_reply]

}



```

#### Expected result:
```
C:\projects\Go\test>go test
2022/02/17 11:16:46
2022/02/17 11:16:46 Reply: 1111
PASS
ok      test    0.058s
```

#### Actual result:
```
C:\projects\Go\test>go test -count 100
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31
2022/02/17 11:05:31 Reply: 1111
2022/02/17 11:05:31 nats: no responders available for request
exit status 1
FAIL    test    0.069s
```

**Comment by nuharaf** [comment]
I think you need flush after subscribe , but even then it probably only work in single server mode. Honestly, I treat subscription like DNS record update, it doesnt happen synchronously across all server.

If I need to do dynamic subscription (with dynamically generated subject), I will simply poll the subscriber using Request until I get some reply.

**Comment by SaulBerrenson** [comment]
> I think you need flush after subscribe , but even then it probably only work in single server mode. Honestly, I treat subscription like DNS record update, it doesnt happen synchronously across all server.
> 
> If I need to do dynamic subscription (with dynamically generated subject), I will simply poll the subscriber using Request until I get some reply.

Man, tnx yoy, it's helped on c client. But it's not idempotent behavior of server, because for problem resolving i need to write here.

**Comment by kozlovic** [comment]
@SaulBerrenson I am guessing by "idempotent behavior" you mean that repeating the test should either always succeed or always fail. But this is an invalid assumption. You are using different connections to subscribe than to send. The server reads from these 2 different connections in different goroutines (see that as threads). So there is no guarantee on which will be processed first in the server, even if in your application you are doing SUB before PUB.
As @nuharaf said, if you were to flush after the sub, and you are using a single server, or are sure that your publisher connects to the same server, then it would always work. If the publish and subscribe use the same connection, then you don't need the flush since then SUB would be processed before the PUB.

In summary, this is expected behavior:
- You have different client connections that subscribe or publish
- Client connections use buffers and a go routine to flush data asynchronously
- Server reads from client connections in different go routine, therefore you can't assume ordering between 2 different client connections

Also, as @nuharaf correctly mentioned, if your SUB connects to server A and your pub connects to server B (in a cluster setup), even doing a flush after SUB would simply guarantee that server A has processed the SUB, but does not guarantee that server B has received the SUB interest. So if you publish to server B even after a SUB flush, then there is no guarantee that the SUB would receive that message.
-
